prisma/schema.prisma

```diff
@@ -23,6 +23,74 @@ datasource db {
   url      = env("DATABASE_URL")
 }
 
+// Achhievement
+// A final achievement can consist of several under achievements
+model achievement_template {
+  id                            Int                       @id @default(autoincrement())
+  metrics                       String[]                  @default([])
+  templateFor                   template_for_enum
+  // Groups defines the "final-achievement"
+  group                         String                    @unique @db.VarChar
+  // Group order tells in which place a subordinate step / achievement is
+  groupOrder                    Int                      
+  // the final achievement is marked with isLast 
+  isLast                        Boolean
+  // what type of achievement is it
+  type                          achievement_type_enum
+  name                          String                    @db.VarChar
+  subtitle                      String                    @db.VarChar
+  description                   String                    @db.VarChar
+  image                         String                    @db.VarChar
+  // some achievements show actions that a user can/must perform next 
+  actionDescription             String?                   @db.VarChar
+  actionRedirectLink            String?                   @db.VarChar
+  actionType                    action_type_enum?
+  // defines what value must be reached to receive the achievement 
+  condition                     String                    @db.VarChar
+  // defined how values must be aggregated to evaluate the condition
+  conditionDataAggregations     Json                      @db.Json
+  isActive                      Boolean 
+}
+model user_achievement {
```

Comment:
Do you mean something like `updateOnAction`? @dhenkel92

Reply:
What I mean is, let's say we have a template and it's already in progress for some users. What do we do if we change the template? Should we:
1) ignore the change for these users.
2) update all user_achievements accordingly.
3) only keep some of the fields updated automatically, like the condition.
4) always create migrations if there was a bug and we want to have it reflected to the user_achievements.

Reply:
Ok I see. 
I find it difficult to answer.  But we need here absolutely still the relation.
Think that it is a decision of LF, what happens when templates change. I just think about: what if the user already has a progress on an achievement and all of a sudden the condition changes? Could be demotivating, also from a pedagogical point of view.

Reply:
In my opinion I think option 1 would be best. We ignore the changes for the previous `user_achievements`. Means it only applies to the achievements that are newly started.

Reply:
I would avoid unnecessary complexity, and after some initial setup I guess we somewhen reach a stable state where achievements rarely change

Reply:
Hm, do we really want to copy all the achievement data _per user_? Couldn't we make the template immutable and reference it via a foreign key here?

Reply:
Also this would mean we could do both easily:
- If something really goes wrong, we could modify the existing template and change stuff for all users
- If we want to make adaptions to how a certain achievement works, we can create a new achievement template with the same name and deactivate the old one

That way it is also easier to keep track of changes (hundreds of achievement templates are easier to keep track of than thousands of user achievements with potentially different conditions etc.)

Reply:
> Hm, do we really want to copy all the achievement data per user? Couldn't we make the template immutable and reference it via a foreign key here? And then duplicate & modify the achievement template whenever it changes?

We thought of this scenario, because copying felt weird to us, too. I think this could add more complexity to maintaining the templates. In your proposal we would need some kind of "versioning" of the templates? So, we know the Template `Y` is the second version of the template `X`.  Maybe this can be solved with a "parent" reference.

Reply:
A "parent" reference would mean we could create a tree of changes. I doubt we ever need (a full version control system) as there is only one active template at one point in time anyways, so a linear history would suffice.

As we already called this "versions" why bot introduce a version Int field? And then make (name, version) a secondary primary key?

Reply:
> We thought of this scenario, because copying felt weird to us, too. I think this could add more complexity to maintaining the templates

Complexity in what way?

Reply:
I rather foresee complexity with the current design:

- Let's say we have 1000 active users, that start 100 achievements each year. If the image field takes 1MB, this means we redundantly store 100GB of data _per year_. I do not see any benefit in that redundancy, but I see the cost of storing this.

- Let's say there is a syntax error in one version of the achievement's condition, after it was changed a couple of times. How do we now fix this, without modifying other versions of that template and changing their logic?

Reply:
> A "parent" reference would mean we could create a tree of changes. I doubt we ever need a full version control system as there is only one active template at one point in time anyways, so a linear history would suffice.

I hear you on that. I don't see any reason why a linear history shouldn't be sufficient. 

> Complexity in what way?

Complexity for the "content team" which creates the achievement in the back office (Retool).
But I guess there's not much more complexity to it. Maybe, we could even add a button in Retool, that duplicates the achievement and let's you edit it, if you want to add a new "version" of an achievement.

Reply:
> As we already called this "versions" why not introduce a version Int field? And then make (name, version) a secondary primary key?

Yeah, let's try this.

Reply:
Okay, so we now use name + increasing id?